docs/docs/render-props.md

```diff
@@ -0,0 +1,224 @@
+---
+id: render-props
+title: Render Props
+permalink: docs/render-props.html
+---
+
+The term ["render prop"](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce) refers to a simple technique for sharing code between React components using a prop whose value is a function.
+
+## Use Render Props for Cross-Cutting Concerns
+
+Components are the primary unit of code reuse in React, but it's not always obvious how to share the state or behavior that one component encapsulates to other components that need that same state.
+
+For example, the following component tracks the mouse position in a web app:
+
+```js
+class MouseTracker extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+        <h1>Move the mouse around!</h1>
+        <p>The current mouse position is ({this.state.x}, {this.state.y})</p>
+      </div>
+    );
+  }
+}
+```
+
+As the cursor moves around the screen, the component displays its (x, y) coordinates in a `<p>`.
+
+Now the question is: How can we reuse this behavior in another component? In other words, if another component needs to know about the cursor position, can we encapsulate that behavior so that we can easily share it with that component?
+
+Since components are the basic unit of code reuse in React, let's try refactoring the code a bit to use a `<Mouse>` component that encapsulates the behavior we need to reuse elsewhere.
+
+```js
+// The <Mouse> component encapsulates the behavior we need...
+class Mouse extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
```

Comment:
Prefix with underscore since it's private (`_handleMouseMove`)? Or is that not really common in open-source?

Reply:
That seems to have fallen out of fashion when using `React`, but I remember using that ll the time. It hated it as I wanted _real_ private methods.

Reply:
I haven't seen it in years.

Reply:
It's pretty common internally at Facebook as we have a custom Babel transform to make them "private" (the names get mangled).

Reply:
@Daniel15 

>  we have a custom Babel transform to make them "actually private"

I WANT! Only I had always dreamed of a transform that took methods and properties that began with an underscore, or prefixed with the `private` keyword, and used [WeakMaps](http://2ality.com/2016/01/private-data-classes.html#keeping-private-data-in-weakmaps) instead of name mangling or Symbols to make them "actually private". Probably an overkill, I know, but still, they would be as close to private that we an get w/ES5. Plus I'm not as much of a purist anymore. I've been beaten down over time. But if I had that custom Babel plugin, maybe it would breath some new life into me. 😘 🙏

Reply:
The old version was open sourced as part of jstransform (https://github.com/facebookarchive/jstransform/blob/aa574cfbf2088e9206b88d700f21794d17f32212/visitors/es6-class-visitors.js#L185-L187), I don't think the Babel version was ever open sourced though :(

Having said that, there's a proposal for private properties so maybe we'll actually have it for realz in JavaScript eventually. https://github.com/tc39/proposal-private-fields